hail/src/main/scala/is/hail/io/gen/ExportBGEN.scala

```diff
@@ -0,0 +1,371 @@
+        swap(left, right - 1)
+        left += 1
+        right -= 1
+      }
+    }
+
+    swap(left, start + n - 1)
+
+    quickSortWithIndex(a, idx, start, left - start)
+    quickSortWithIndex(a, idx, left + 1, n - (left - start + 1))
+  }
+
+  def sortedIndex(a: Array[Double], idx: Array[Int]) {
+    val n = a.length
+    assert(idx.length == n)
+    resetIndex(idx)
+    quickSortWithIndex(a, idx, 0, n)
+  }
+
+  def roundWithConstantSum(input: Array[Double], fractional: Array[Double], index: Array[Int],
+    indexInverse: Array[Int], output: ArrayBuilder[Byte], expectedSize: Long) {
+    val n = input.length
+    assert(fractional.length == n && index.length == n && indexInverse.length == n)
+
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i)
+      val f = x - x.floor
+      fractional(i) = f
+      totalFractional += f
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+
+    sortedIndex(fractional, index)
+
+    i = 0
+    while (i < n) {
+      indexInverse(index(i)) = i
+      i += 1
+    }
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (indexInverse(i) < F) input(i).ceil.toInt else input(i).floor.toInt
+      assert(r >= 0 && r < 256)
+      if (i != n - 1)
+        output += r.toByte
+      newSize += r
+      i += 1
+    }
+    assert(newSize == expectedSize)
+  }
+
+
+  private def emitGPData(nAlleles: Int) {
+    uncompressedData.clear()
+    val nGenotypes = triangle(nAlleles)
+
+    intToBytesLE(uncompressedData, nSamples)
+    shortToBytesLE(uncompressedData, nAlleles)
+    uncompressedData += ploidy
+    uncompressedData += ploidy
+
+    val gpResized = new Array[Double](nGenotypes)
+    val index = new Array[Int](nGenotypes)
+    val indexInverse = new Array[Int](nGenotypes)
+    val fractional = new Array[Double](nGenotypes)
+
+    val samplePloidyStart = uncompressedData.length
+    var i = 0
+    while (i < nSamples) {
+      uncompressedData += 0x82.toByte // placeholder for sample ploidy - default is missing
+      i += 1
+    }
+
+    uncompressedData += phased
+    uncompressedData += 8.toByte
+
+    def emitNullGP() {
+      var gIdx = 0
+      while (gIdx < nGenotypes - 1) {
+        uncompressedData += 0
+        gIdx += 1
+      }
+    }
+
+    i = 0
+    while (i < nSamples) {
+      gs.setGenotype(i)
+
+      if (gs.hasGP) {
+        var idx = 0
+        var gpSum = 0d
+        while (idx < nGenotypes) {
+          val x = gs.getGP(idx)
+          gpSum += x
+          gpResized(idx) = x * totalProb // Assuming sum(GP) == 1
+          idx += 1
+        }
+
+        if (gpSum >= 0.999 && gpSum <= 1.001) {
+          uncompressedData(samplePloidyStart + i) = ploidy
+          roundWithConstantSum(gpResized, fractional, index, indexInverse, uncompressedData, totalProb)
+        } else {
+          warn(s"GP sum was not in the range [0.999, 1.001]. Found $gpSum. Emitting missing probabilities.")
```

Comment:
this warning will go to the worker logs, creating massive logs if something is wrong in the data but not propagating back to the user. Prefer an incremented count of bad GPs per partition which is collected with the partition file name.